### ISSUE TITLE
(*pixel.Batch).MakePicture: Picture is not the Batch's Picture

### ISSUE THREAD

**Issue by rumd3x** [faiface/pixel#187]
Hello,

I've started a little project. The code at this point is basically a rip of the platformer example, but structured in different packages, and built around some interfaces, so I can expand on it.

The problem is that whenever i try to draw to an IMDraw, the window is created for a split second, and then closes, with the following error: `panic: (*pixel.Batch).MakePicture: Picture is not the Batch's Picture`. I can't seem to find out why.

Also, if I comment out the following snippet of code at `MakePicture@batch.go:83` the error is obviously not thrown, and the window runs as normal, but stays blank:
```go
if p != b.cont.Picture {
	panic(fmt.Errorf("(%T).MakePicture: Picture is not the Batch's Picture", b))
}
```

Here are two snippets of code that throws this error when calling the `Draw` method:

```go
//Text implements entity and represents a string on the canvas
type Text struct {
	Txt    string
	TTF    *text.Text
	ZIndex int
}

//NewText creates a new Text struct
func NewText(txt string, origin pixel.Vec, size float64, font string) *Text {
	font = fmt.Sprintf("%s.ttf", font)
	self := utils.LoadTTF(font, size, origin)
	return &Text{
		Txt:    txt,
		TTF:    self,
		ZIndex: 10,
	}
}

//Draw draws the entity on the target
func (t *Text) Draw(imd *imdraw.IMDraw) {
	t.TTF.WriteString(t.Txt)
	t.Draw(imd)
}
```

```go
//Player is the player and implements entity
type Player struct {
	dir              float64
	weight           float64
	speedStep        float64
	sprintMultiplier float64
	maxVelocity      pixel.Vec
	currentVelocity  pixel.Vec
	position         pixel.Rect
	anim             *playerAnimation
}

type playerAnimation struct {
	state       int
	counter     float64
	rate        float64
	frame       pixel.Rect
	sprite      pixel.Sprite
	spritesheet pixel.Picture
	anims       map[string][]pixel.Rect
	ZIndex      int
}

//NewPlayer creates a new player
func NewPlayer(pos pixel.Vec) *Player {
	sheet, anims, _ := utils.LoadAnimationSheet("player.png", "player.csv", 12)

	sprite := pixel.NewSprite(nil, pixel.Rect{})

	anim := playerAnimation{
		state:       0,
		counter:     0,
		sprite:      *sprite,
		frame:       anims["Front"][0],
		anims:       anims,
		spritesheet: sheet,
		rate:        1.0 / 10,
		ZIndex:      1000,
	}

	rect := pixel.R(pos.X, pos.Y, pos.X+12, pos.Y-12)
	return &Player{
		dir:              1,
		weight:           50,
		speedStep:        5,
		sprintMultiplier: 2,
		maxVelocity:      pixel.V(100, 300),
		currentVelocity:  pixel.V(0, 0),
		position:         rect,
		anim:             &anim,
	}
}

//Draw draws the entity on the target
func (p *Player) Draw(imd *imdraw.IMDraw) {
	p.anim.sprite.Set(p.anim.spritesheet, p.anim.frame)
	p.anim.sprite.Draw(imd,
		pixel.IM.ScaledXY(pixel.ZV, pixel.V(
			p.position.W()/p.anim.sprite.Frame().W(),
			p.position.H()/p.anim.sprite.Frame().H(),
		)).ScaledXY(pixel.ZV, pixel.V(-p.dir, 1)),
	)
}
```

So, what am I missing? Thanks in advance.

**Comment by faiface** [comment]
Hi! This is because a single Batch or IMDraw (IMDraw uses a Batch under the hood) can only be used with a single Picture as described [here](https://godoc.org/github.com/faiface/pixel#Batch) and [here](https://godoc.org/github.com/faiface/pixel/imdraw#New).

All your animation frames are from the same Picture because they are from the same sheet. Thus you can all draw them to the same Batch or IMDraw and it will work.

However, the Text object uses a different Picture: it uses a pre-rendered atlas of glyphs. You can obtain that Picture by [Atlas.Picture](https://godoc.org/github.com/faiface/pixel/text#Atlas.Picture/) but you need to use another Batch or IMDraw. Or not use any.

Btw, if you Draw function doesn't use any functionality specific to IMDraw, it's best to accept a generic [pixel.Target](https://godoc.org/github.com/faiface/pixel#Target). That way you are free to pass whatever you want from the outside.

**Comment by rumd3x** [comment]
Thank you for the quick response.

I never tried to draw them both at the same time, and the IMDraw as initialized with nil. That works fine on the platformer example. So I didn't fully understood why I was getting the error.

However, your answer was very helpful, as I probably was not using the IMDraw the intended way, and also had searched for something like the Atlas.Picture, but was unable to find it. So I changed the interface contract slightly, and was able to get it working.

Here's what I ended up with:
```go
//Text implements entity and represents a string on the canvas
type Text struct {
	TTF     *text.Text
	ZIndex  int
	Picture pixel.Picture
}

//NewText creates a new Text struct
func NewText(txt string, origin pixel.Vec, size float64, font string) *Text {
	font = fmt.Sprintf("%s.ttf", font)

	ttf := utils.LoadTTF(font, size, origin)
	ttf.WriteString(txt)
	ttf.Color = colornames.White

	return &Text{
		TTF:     ttf,
		ZIndex:  10,
		Picture: ttf.Atlas().Picture(),
	}
}

//Draw draws the entity on the target
func (t *Text) Draw() *imdraw.IMDraw {
	imd := imdraw.New(t.Picture)
	imd.Precision = 32

	t.TTF.Draw(imd, pixel.IM)
	return imd
}
```

```go
//Player is the player and implements entity
type Player struct {
	dir              float64
	weight           float64
	speedStep        float64
	sprintMultiplier float64
	maxVelocity      pixel.Vec
	currentVelocity  pixel.Vec
	position         pixel.Rect
	anim             *playerAnimation
}

type playerAnimation struct {
	state       int
	counter     float64
	rate        float64
	frame       pixel.Rect
	sprite      pixel.Sprite
	spritesheet pixel.Picture
	anims       map[string][]pixel.Rect
	ZIndex      int
}

//NewPlayer creates a new player
func NewPlayer(pos pixel.Vec) *Player {
	sheet, anims, _ := utils.LoadAnimationSheet("player.png", "player.csv", 12)

	sprite := pixel.NewSprite(nil, pixel.Rect{})

	anim := playerAnimation{
		state:       0,
		counter:     0,
		sprite:      *sprite,
		frame:       anims["Front"][0],
		anims:       anims,
		spritesheet: sheet,
		rate:        1.0 / 10,
		ZIndex:      1000,
	}

	rect := pixel.R(pos.X, pos.Y, pos.X+12, pos.Y-12)
	return &Player{
		dir:              1,
		weight:           50,
		speedStep:        5,
		sprintMultiplier: 2,
		maxVelocity:      pixel.V(100, 300),
		currentVelocity:  pixel.V(0, 0),
		position:         rect,
		anim:             &anim,
	}
}

//Draw draws the entity on the target
func (p *Player) Draw() *imdraw.IMDraw {
	imd := imdraw.New(p.anim.spritesheet)
	imd.Precision = 32

	fmt.Printf("Position: %+v\n", p.position)

	p.anim.sprite.Set(p.anim.spritesheet, p.anim.frame)
	p.anim.sprite.Draw(imd,
		pixel.IM.ScaledXY(pixel.ZV, pixel.V(
			p.position.W()/p.anim.sprite.Frame().W(),
			p.position.H()/p.anim.sprite.Frame().H(),
		)).ScaledXY(pixel.ZV, pixel.V(-p.dir, -1)).
			Moved(p.position.Center()),
	)

	return imd
}
```

Thank you for your help, time and patience, and also for this amazing library.

**Comment by faiface** [comment]
> and the IMDraw as initialized with nil.

Yeah, that was the problem then. The IMDraw has to be initialized to the same picture as what you're drawing on it. In the platformer example, the guy is never drawn on the IMDraw.

> searched for something like the Atlas.Picture, but was unable to find it

> `Picture: ttf.Atlas().Picture(),`

That's it! That's Atlas.Picture (in my previous post, I included a link to the doc). When I wrote Atlas.Picture, Atlas was referring to a type.

I'm happy you got it working, now you have something to move on from. However, what you ended up with is quite inefficient, because you are constructing new and new IMDraws every single time.

So:
1. There's no need to use IMDraw with Text unless you want to somehow combine it with some primitive graphics.
2. I'd suggest changing the signatures of the Draw methods to: `Draw(target pixel.Taget)`. Target is an interface and IMDraw is a Target as well as pixelgl.Window. Then I'd simply create one Batch or an IMDraw outside of the Draw methods so that you don't have to refcreate it every time, draw the player there and draw the text directly to the window.

Btw, if you haven't seen the tutorial, I'd highly recommend you check it out.
